### PR TITLE
Add TransformsIntoTransforms to enable queued MCV redeploy.

### LINKED ISSUE #17602: Allow conyard to accept and queue mcv's order.
some people experience "conyard deployment issue"
As you can now make a conyard undeploy with a force move order,
and shift queue a deploy order
I guess people started to do the following sequence : force move + shift + f
The problem is :
Conyard won't accept the shift F before being in mcv form
So the player thinks the deploy order is queued but it is not.

I would like to improve the new ALT+MOVE for conyard by allowing the conyard to queue any order an mcv can take.

### PR DESCRIPTION
**Pull request by pchote**
Fixes #17602.

### CODE REVIEW

**Comment by pchote** [comment]
This has been sitting open on the milestone with no activity for quite a while now. Can we please get some movement here, even if that happens to be an "I don't think we should take this for the playtest"?

**Comment by Mailaender** [comment]
Sorry, I can't test this as the ALT key is somehow blocked in KDE even in full screen mode and the conyard won't undeploy at all. See https://github.com/OpenRA/OpenRA/issues/15879 for making it a remappable key.

**Comment by pchote** [comment]
@Mailaender try https://bbs.archlinux.org/viewtopic.php?id=109085

**Review by Mailaender — DISMISSED** [review]
Works as promised.

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		[Desc("Require the force-move modifier to display the enter cursor.")]
+		public readonly bool RequiresForceMove = false;
+
+		public override object Create(ActorInitializer init) { return new TransformsIntoTransforms(init.Self, this); }
+	}
+
+	public class TransformsIntoTransforms : ConditionalTrait<TransformsIntoTransformsInfo>, IResolveOrder, IOrderVoice, IIssueDeployOrder
+	{
+		readonly Actor self;
+		Transforms[] transforms;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs]: Not needed.

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Add to a building to allow queued transform orders while undeploying.")]
+	public class TransformsIntoTransformsInfo : ConditionalTraitInfo, Requires<TransformsInfo>, Requires<IHealthInfo>
+	{
+		[VoiceReference]
+		public readonly string Voice = "Action";
+
+		[Desc("Require the force-move modifier to display the enter cursor.")]
+		public readonly bool RequiresForceMove = false;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs]: Not used.

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Activities;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs]: Not used.

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs]: Not used.

File: OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+			: base(info)
+		{
+			this.self = self;
+		}
+
+		protected override void Created(Actor self)
+		{
+			transforms = self.TraitsImplementing<Transforms>().ToArray();
+			base.Created(self);
+		}
```
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Buildings/TransformsIntoTransforms.cs]: Not used after the change above.

**Review by teinarss — CHANGES_REQUESTED** [review]

**Comment by pchote** [comment]
Fixed (first and third pushes) and rebased (second push).

**Review by abcdefg30 — DISMISSED** [review]
Works as advertised. One thing however: When undeploying you can now issue deploy orders. However, when deploying you can also issue deploy orders (to immediately undeploy again) which are not executed. Not supporting immediate undeploy would not be an issue imho if the command bar icon for deploying was disabled.

**Comment by pchote** [comment]
Hmm... this may require more thought then. I had always considered blocking immediate undeploy as a protection against accidental double keypresses so its not clear whether we want to support that. Keeping the current assymetric state of the PR also isn't great.

Would it be confusing if we only allowed this queued redeploy if the actor has moved to a different location?

**Comment by pchote** [comment]
> However, when deploying you can also issue deploy orders (to immediately undeploy again) which are not executed.

Are you sure about this? https://github.com/OpenRA/OpenRA/pull/17615/files#diff-77ec43bbb0b6febb3e3e53220fef3002R37-R42 is supposed to make sure that you can only issue deploy orders when *un*deploying - the initial deploy animation should be handled by the make animation rather than Transforms.

**Comment by WhoCaresora** [comment]
I don't see a single case where queuing order when deploying is usefull as the only thing a conyard can do is undeploy. 

The reasons that come in mind to deploy and mcv : 

- To have build radius and by extension placing a building. -- Queuing immediate undeploy would not give the necessary time to place a single structure
- To have additional construction speed -- You want your mcv as conyard form until it is needed somewhere so again queuing a undeploying order does not apply
- changing mcv armor into conyard concrete -- You might want your mcv in conyard until the zone is safe so undeploying immediatly put you back in a vulnerable position.
- Escaping a hijacker. I guess with hijacker you could deploy-undeploy to make his targeting void as he can't target building and hope your opponent to not be attentive (but seriously who would watch somewhere else when he is about to capture and mcv).
-escaping any unit with opportunity fire. Unit with opportunity fire do not transfer the focus fire over transformation. you can make planes "bug" big time and you can flee from tanks.

**Comment by PunkPun** [comment]
> Hmm... this may require more thought then. I had always considered blocking immediate undeploy as a protection against accidental double keypresses so its not clear whether we want to support that. Keeping the current assymetric state of the PR also isn't great.
> 
> Would it be confusing if we only allowed this queued redeploy if the actor has moved to a different location?

A normal deploy order while deploying should definitely not queue an undeploy. If it was a case it would become a source of frustration to many people. However allowing shift-f to queue an undeploy should be done for the sake of consistency and player expectations. Though it is not necessary to do now as there are no ingame incentives for this action apart from providing temporary construction radius (what would be clunky and rarely the desired action)

**Comment by pchote** [comment]
Fixed. The commandbar button now stays disabled unless shift is held to queue the redeploy.

**Review by teinarss — APPROVED** [review]

**Comment by pchote** [comment]
I count two :+1:s here. @abcdefg30 did you want to take another look here or shall I merge this?

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a03d7bc235b3cf7d8fcb0588fbb190508c740dd0...824d9682500c26697974aaf185a2b7ad45f98877)

**Pull request merged**